File: solutions/8. String to Integer (atoi)/Python/String_to_Integer_(atoi).py
class Solution:
    def myAtoi(self, s: str) -> int:
        """
            state = 0: terminate
            state = 1: wait
            state = 2: read
        """
        state = 1
        stringLength = len(s)
        reader = 0
        signBit = 1
        number = 0

        while state and reader < stringLength: #do while not in terminate state (state = 0)
            character = s[reader]

            if state == 1:
                if character == ' ':
                    state = 1

                elif character == '+' or character == '-':
                    state = 2
                    if character == '-':
                        signBit = -1

                elif character >= '0' and character <= '9':
                    state = 2
                    number = int(character)
                
                else:
                    state = 0

            elif state == 2:
                if character >= '0' and character <= '9':
                    state = 2
                    number = number * 10 + int(character)
                
                else:
                    state = 0

            else:
                print("error")

            reader = reader + 1


        #int in python will never overflow, thus we can calculate the final resault afterwards
        number = signBit * number

        if number > 2147483647:
            number = 2147483647
        elif number < -2147483648:
            number = -2147483648

        return number

S = Solution()
print(S.myAtoi("-91"))

File: Solutions/8. String to Integer (atoi)/CPP/String_to_Integer_(atoi).cpp
#include <iostream>

using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        /*
            state = 0: terminate
            state = 1: wait
            state = 2: read
        */
        int state = 1;
        int stringLength = s.length();
        int reader = 0;
        int signBit = 1; //set default to '+'
        long outputNumber = 0;

        while(state && reader < stringLength) {
            char character = s[reader];
            switch(state) {
                case 1:
                    if(character == ' ') {
                        //state = 1; state remain the same
                    }
                    else if(character == '+' || character == '-') {
                        state = 2;
                        if(character == '-') {
                            signBit = -1;
                        }
                    }
                    else if(character >= '0' && character <= '9') {
                        state = 2;
                        outputNumber = int(character) - '0'; //transfer the ascii code to integer, and store it into the outputNumber
                    }
                    else { //get invalid character
                        state = 0;
                    }
                    break;

                case 2:
                    if(character >= '0' && character <= '9') {
                        //state = 3; state remain the same

                        //overflow detection
                        outputNumber = outputNumber  *10 + signBit  *(int(character) - '0');
                        if(outputNumber > 2147483647) {
                            outputNumber = 2147483647;
                            state = 0;
                        }
                        else if(outputNumber < -2147483648) {
                            outputNumber = -2147483648;
                            state = 0;
                        }
                    }
                    else { //get invalid character
                        state = 0;
                    }
                    break;

                default:
                    cout << "error" << endl;
                    break;
            }
            reader = reader + 1; //move the reader to the next character
        }

        return int(outputNumber);
    }
};

int main() {
    Solution S;

    //input
    string s = "-21474836";

    cout << S.myAtoi(s) << endl;

    return 0;
}
